bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/library/types/QuantityType.java

```diff
@@ -0,0 +1,218 @@
+                return false;
+            }
+        } else if (quantity.getUnit().getDimension() != other.quantity.getUnit().getDimension()) {
+            return false;
+        } else if (compareTo(other) != 0) {
+            return false;
+        }
+
+        return true;
+    }
+
+    @Override
+    public int compareTo(QuantityType o) {
+        if (quantity.getUnit().isCompatible(o.quantity.getUnit())) {
+            QuantityType v1 = this.toUnit(getUnit().getSystemUnit());
+            QuantityType v2 = o.toUnit(o.getUnit().getSystemUnit());
+            return Double.compare(v1.doubleValue(), v2.doubleValue());
+        } else {
+            throw new IllegalArgumentException("Can not compare incompatible units");
+        }
+    }
+
+    public Unit<?> getUnit() {
+        return quantity.getUnit();
+    }
+
+    public Dimension getDimension() {
+        return getUnit().getDimension();
+    }
+
+    @SuppressWarnings({ "unchecked", "rawtypes" })
+    public QuantityType toUnit(Unit targetUnit) {
+        try {
+            if (targetUnit != getUnit()) {
+                UnitConverter uc = getUnit().getConverterTo(targetUnit);
+                Quantity result = Quantities.getQuantity(uc.convert(quantity.getValue()), targetUnit);
+
+                return new QuantityType(result.getValue().doubleValue(), targetUnit);
+            } else {
+                return this;
+            }
+        } catch (UnconvertibleException e) {
+            return null;
+        }
+    }
+
+    public QuantityType toUnit(String targetUnit) {
+        if (targetUnit != null) {
+            Unit<?> unit = AbstractUnit.parse(targetUnit);
+            if (unit != null) {
+                return toUnit(unit);
+            }
+
+        }
+        return null;
+    }
+
+    public BigDecimal toBigDecimal() {
+        return new BigDecimal(doubleValue());
+    }
+
+    @Override
+    public int hashCode() {
+        final int prime = 31;
+        int tmp = prime * (getUnit() == null ? 0 : getUnit().hashCode());
+        tmp += prime * (quantity.getValue() == null ? 0 : quantity.getValue().hashCode());
+        return tmp;
+    }
+
+    @Override
+    public String format(String pattern) {
+        return String.format(pattern, quantity.getValue());
+    }
+
+    @Override
+    public int intValue() {
+        return quantity.getValue().intValue();
+    }
+
+    @Override
+    public long longValue() {
+        return quantity.getValue().longValue();
+    }
+
+    @Override
+    public float floatValue() {
+        return quantity.getValue().floatValue();
+    }
+
+    @Override
+    public double doubleValue() {
+        return quantity.getValue().doubleValue();
+    }
+
+    @Override
+    public String toFullString() {
+        return quantity.toString();
+    }
+
+    @Override
+    public State as(Class<? extends State> target) {
+        if (target == OnOffType.class) {
+            return intValue() == 0 ? OnOffType.OFF : OnOffType.ON;
```

Comment:
I assume it doesn't care, but shouldn't we use the double value, so 0.5 will be converted to "ON", too?